// @convex-dev/better-auth/dist/auth.js
import { betterAuth } from "better-auth";
import { anonymous, bearer, emailOTP, genericOAuth, jwt, magicLink, oidcProvider, oneTap, oneTimeToken, phoneNumber, twoFactor, username, } from "better-auth/plugins";
import { convex } from "./plugins/convex/index.js";
import { passkey } from "better-auth/plugins/passkey";
import { convexAdapter } from "./client/adapter.js";
// This is the config used to generate the schema
const options = {
    logger: {
        disabled: true,
    },
    database: convexAdapter({}, {}),
    rateLimit: {
        storage: "database",
    },
    plugins: [
        twoFactor(),
        anonymous(),
        username(),
        phoneNumber(),
        magicLink({ sendMagicLink: async () => { } }),
        emailOTP({ sendVerificationOTP: async () => { } }),
        passkey(),
        genericOAuth({
            config: [
                {
                    clientId: "",
                    clientSecret: "",
                    providerId: "",
                },
            ],
        }),
        oneTap(),
        oidcProvider({
            loginPage: "/login",
        }),
        bearer(),
        oneTimeToken(),
        jwt(),
        convex(),
    ],
}; // assert type to avoid overloading ts compiler
const config = betterAuth(options);
export { config as auth };
