// @convex-dev/better-auth/dist/react-start/index.js
import { betterAuth } from "better-auth";
import { createCookieGetter } from "better-auth/cookies";
import { betterFetch } from "@better-fetch/fetch";
import * as jose from "jose";
import {} from "convex/server";
import { JWT_COOKIE_NAME } from "../plugins/convex/index.js";
import { ConvexHttpClient } from "convex/browser";
import { getStaticAuth } from "../client/index.js";
export const getCookieName = (createAuth) => {
    const createCookie = createCookieGetter(getStaticAuth(createAuth).options);
    const cookie = createCookie(JWT_COOKIE_NAME);
    return cookie.name;
};
export const getCookieNames = (createAuth) => {
    const createCookie = createCookieGetter(getStaticAuth(createAuth).options);
    return {
        convexJwt: createCookie(JWT_COOKIE_NAME).name,
        sessionToken: createCookie("session_token").name,
    };
};
export const setupFetchClient = async (createAuth, getCookie) => {
    const createClient = () => {
        const sessionCookieName = getCookieName(createAuth);
        const token = getCookie(sessionCookieName);
        const client = new ConvexHttpClient(process.env.VITE_CONVEX_URL);
        if (token) {
            client.setAuth(token);
        }
        return client;
    };
    return {
        fetchQuery(query, args) {
            return createClient().query(query, args);
        },
        fetchMutation(mutation, args) {
            return createClient().mutation(mutation, args);
        },
        fetchAction(action, args) {
            return createClient().action(action, args);
        },
    };
};
export const fetchSession = async (request, opts) => {
    if (!request) {
        throw new Error("No request found");
    }
    const convexSiteUrl = opts?.convexSiteUrl ?? process.env.VITE_CONVEX_SITE_URL;
    if (!convexSiteUrl) {
        throw new Error("VITE_CONVEX_SITE_URL is not set");
    }
    const { data: session } = await betterFetch("/api/auth/get-session", {
        baseURL: convexSiteUrl,
        headers: {
            cookie: request.headers.get("cookie") ?? "",
        },
    });
    return {
        session,
    };
};
export const fetchAuth = async (request, opts) => {
    if (!request) {
        throw new Error("No request found");
    }
    const convexSiteUrl = opts?.convexSiteUrl ?? process.env.VITE_CONVEX_SITE_URL;
    if (!convexSiteUrl) {
        throw new Error("VITE_CONVEX_SITE_URL is not set");
    }
    const { data } = await betterFetch("/api/auth/convex/token", {
        baseURL: convexSiteUrl,
        headers: {
            cookie: request.headers.get("cookie") ?? "",
        },
    });
    if (!data?.token) {
        return;
    }
    const claims = jose.decodeJwt(data.token);
    return {
        token: data.token,
        userId: claims.sub,
    };
};
export const getAuthFromCookie = (cookie, { tolerance = 10 } = {}) => {
    if (!cookie) {
        return;
    }
    const claims = jose.decodeJwt(cookie);
    const exp = claims?.exp;
    const now = Math.floor(new Date().getTime() / 1000);
    const isExpired = exp ? now > exp + tolerance : true;
    if (isExpired) {
        return;
    }
    return { userId: claims?.sub, token: cookie };
};
export const getAuth = async (request, getCookie, createAuth, opts) => {
    const sessionCookieName = getCookieName(createAuth);
    const token = getCookie(sessionCookieName);
    const { session } = await fetchSession(request, opts);
    return {
        userId: session?.user.id,
        token,
    };
};
export const reactStartHandler = (request, opts) => {
    const requestUrl = new URL(request.url);
    const convexSiteUrl = opts?.convexSiteUrl ?? process.env.VITE_CONVEX_SITE_URL;
    if (!convexSiteUrl) {
        throw new Error("VITE_CONVEX_SITE_URL is not set");
    }
    const nextUrl = `${convexSiteUrl}${requestUrl.pathname}${requestUrl.search}`;
    const headers = new Headers(request.headers);
    headers.set("accept-encoding", "application/json");
    headers.set("host", convexSiteUrl);
    return fetch(nextUrl, {
        method: request.method,
        headers,
        redirect: "manual",
        body: request.body,
        // @ts-expect-error - duplex is required for streaming request bodies in modern fetch
        duplex: "half",
    });
};
